### PR TITLE
support denoise , sharpening and deinterlace.

### PR DESCRIPTION
**Pull request by xuguangxin**
This is a code refact for https://github.com/01org/libyami/pull/463

### CODE REVIEW

**Comment by xuguangxin** [comment]
@mypopydev , 
Please help to review, it's meet your requirements or not.

File: vpp/vaapipostprocess_scaler.cpp [hunk 1]
```diff
@@ -18,12 +18,15 @@
 #endif
 
 #include "vaapipostprocess_scaler.h"
```
Comment (mypopydev) [inline vpp/vaapipostprocess_scaler.cpp]: Put the denosie/de-interlance/... in vaapipostprocess_scaler.cpp/vaapipostprocess_scaler.h maybe is not good idea.

File: vpp/vaapipostprocess_scaler.cpp [hunk 2]
```diff
@@ -59,9 +62,29 @@ VAProcColorStandardType fourccToColorStandard(uint32_t fourcc)
     return VAProcColorStandardBT601;
 }
 
+VaapiPostProcessScaler::VaapiPostProcessScaler()
```
Comment (mypopydev) [inline vpp/vaapipostprocess_scaler.cpp:65]: Put the denosie/de-interlace/... in vaapipostprocess_scaler.cpp/vaapipostprocess_scaler.h maybe is not good idea.

File: vpp/vaapipostprocess_base.cpp [hunk 3]
```diff
@@ -70,6 +65,28 @@ void VaapiPostProcessBase::cleanupVA()
     m_display.reset();
 }
 
+YamiStatus VaapiPostProcessBase::queryVideoProcFilterCaps(
+    VAProcFilterType filterType, void* filterCaps, uint32_t* numFilterCaps)
+{
+    if (!filterCaps)
```
Comment (mypopydev) [inline vpp/vaapipostprocess_base.cpp:71]: !filterCaps && !numFilterCaps more better

File: vpp/vaapipostprocess_scaler.cpp (hunk 2 above)
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp:65]: It's same process. It share most of code and our driver can do deinterlace and scale in same time.

File: vpp/vaapipostprocess_base.cpp (hunk 3 above)
Comment (xuguangxin) [inline vpp/vaapipostprocess_base.cpp:71]: The default value for numFilterCaps is null, it's means we only have one caps, we do not care about the return value.But you are right, i need init tmp to 1 for this case.

**Comment by mypopydev** [comment]
As my view, when call this function, the caller used filterCaps == NULL && numFilterCaps == NULL is wrong, andI think you can't handle this case in the current code  :)

But filterCaps != NULL && numFilterCaps == NULL  or filterCaps == NULL && numFilterCaps != NULL is Ok for this function.

File: vpp/vaapipostprocess_scaler.cpp [hunk 4]
```diff
@@ -85,10 +108,173 @@ VaapiPostProcessScaler::process(const SharedPtr<VideoFrame>& src,
...
+            return false;
+        }
+    }
+    return mapToRange(value, caps->range.min_value, caps->range.max_value, level, minLevel, maxLevel);
+}
+
+YamiStatus
+VaapiPostProcessScaler::setParamToNone(ProcParams& params, int32_t none)
+{
+    params.level = none;
```
Comment (mypopydev) [inline vpp/vaapipostprocess_scaler.cpp:156]: What's the means for the "none“? and the function name is setParamToNone() :)

File: vpp/vaapipostprocess_base.cpp (hunk 3 above)
Comment (mypopydev) [inline vpp/vaapipostprocess_base.cpp:71]: As my view, when call this function, the caller used filterCaps == NULL && numFilterCaps == NULL is wrong, and I think you can't handle this case in the current code :)

BTW: I think filterCaps != NULL && numFilterCaps == NULL or filterCaps == NULL && numFilterCaps != NULL is Ok when call this function.

File: vpp/vaapipostprocess_scaler.cpp (hunk 4 above)
Comment (xuguangxin) [inline vpp/vaapipostprocess_scaler.cpp:156]: none means do not do any denoise and sharpening.

File: vpp/vaapipostprocess_base.cpp (hunk 3 above)
Comment (xuguangxin) [inline vpp/vaapipostprocess_base.cpp:71]: no, current no user case for filterCaps == NULL && numFilterCaps != NULL, all caller will provide filterCaps. I do not want this function too complicate

Comment (mypopydev) [inline vpp/vaapipostprocess_base.cpp:71]: suppose have a case: the caller used 2 steps to get all Caps:
step1: Used  filterCaps == NULL && numFilterCaps != NULL to get accurate number of filter Caps.
       uint32_t numFilterCaps;
       queryVideoProcFilterCaps(filterType, NULL, &numFilterCaps);
step 2: 
      filterCaps = malloc(numFilterCaps \* sizeof(xxx));
      queryVideoProcFilterCaps(filterType,  filterCaps,  NULL);

if not, the caller maybe always use this function like this:

 filterCaps [MAX_NUM_FILTER_CAPS];
 queryVideoProcFilterCaps(filterType,  filterCaps, &numFilterCaps);

 It is not a friendly interface to the caller if can't support the filterCaps == NULL && numFilterCaps != NULL case. :)

Comment (xuguangxin) [inline vpp/vaapipostprocess_base.cpp:71]: let me tell you the truth.
I965 driver do not do not support null filter_caps.

```
VAStatus i965_QueryVideoProcFilterCaps(
    VADriverContextP    ctx,
    VAContextID         context,
    VAProcFilterType    type,
    void               *filter_caps,
    unsigned int       *num_filter_caps
    )
{
    unsigned int i = 0;
    struct i965_driver_data *const i965 = i965_driver_data(ctx);

    if (!filter_caps || !num_filter_caps)
        return VA_STATUS_ERROR_INVALID_PARAMETER;
```

Comment (mypopydev) [inline vpp/vaapipostprocess_base.cpp:71]: oops

**Comment by xuguangxin** [comment]
Three driver issue will expose by this patch.
1. driver do not support denoise from i420 to nv12.
   
   It will crash in driver.This already fixed by Haihao
2. sharpening from i420 to nv12 will report unsupported.
   
   report to libva as https://bugs.freedesktop.org/show_bug.cgi?id=96987
3.  from nv12 to nv12, sharpening + scale will not get right behaviors.
   
   report to libva as https://bugs.freedesktop.org/show_bug.cgi?id=96988

**Comment by seanvk** [comment]
Thanks @xuguangxin for the bugs filed against the driver.

**Comment by xuguangxin** [comment]
@seanvk, It's my pleasure

**Comment by xuguangxin** [comment]
Hi @mypopydev 
Please +1 for me if you think it's ok.

**Comment by mypopydev** [comment]
+1

**Comment by xuguangxin** [comment]
@wudping , Please help verify this, make sure vpp/unittest all passed.
thanks

**Comment by wudping** [comment]
okay, Guangxin.

**Comment by wudping** [comment]
verify + 1

[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from VaapiPostProcessScalerTest
[ RUN      ] VaapiPostProcessScalerTest.Factory
[       OK ] VaapiPostProcessScalerTest.Factory (0 ms)
[ RUN      ] VaapiPostProcessScalerTest.mapToRange
[       OK ] VaapiPostProcessScalerTest.mapToRange (10 ms)
[ RUN      ] VaapiPostProcessScalerTest.Denoise
[       OK ] VaapiPostProcessScalerTest.Denoise (59 ms)
[ RUN      ] VaapiPostProcessScalerTest.Sharpening
[       OK ] VaapiPostProcessScalerTest.Sharpening (41 ms)
[ RUN      ] VaapiPostProcessScalerTest.Deinterlace
[       OK ] VaapiPostProcessScalerTest.Deinterlace (59 ms)
[----------] 5 tests from VaapiPostProcessScalerTest (170 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (171 ms total)
[  PASSED  ] 5 tests.

**Comment by xuguangxin** [comment]
merged as f6f1483a62dfd1424483ecb24ec9305ea96c2093

**Pull request closed without merging**